### PR TITLE
RavenDB-16117 Database Storage Report: Combine items to Grouped Data

### PR DESCRIPTION
**Pull request by Danielle9897**

### CODE REVIEW

File: src/Raven.Studio/typescript/models/database/status/storageReportItem.ts [hunk 1]
```diff
@@ -8,6 +8,7 @@ class storageReportItem {
     type: string;
     internalChildren: storageReportItem[];
     size?: number;
+    length?: number;
```
Comment (ml054) [inline src/Raven.Studio/typescript/models/database/status/storageReportItem.ts:11]: what this length represents?

File: src/Raven.Studio/typescript/viewmodels/database/status/storageReport.ts [hunk 2]
```diff
@@ -18,6 +18,7 @@ type positionAndSizes = {
 class storageReport extends viewModelBase {
 
     static readonly animationLength = 200;
+    static readonly maxChildrenToShow = 100; // todo - make this 1000 ?
```
Comment (ml054) [inline src/Raven.Studio/typescript/viewmodels/database/status/storageReport.ts]: yes

File: src/Raven.Studio/typescript/viewmodels/database/status/storageReport.ts [hunk 3]
```diff
@@ -172,28 +174,74 @@ class storageReport extends viewModelBase {
         };
         return preAllocatedBuffers;
     }
+    
+    private groupChildren(items: storageReportItem[]): storageReportItem[] {
+        
+        if (items.length > storageReport.maxChildrenToShow) {
+            let clonedItems = _.cloneDeep(items);
```
Comment (ml054) [inline src/Raven.Studio/typescript/viewmodels/database/status/storageReport.ts]: why you deep clone here? Can we avoid that?

File: src/Raven.Studio/typescript/viewmodels/database/status/storageReport.ts [hunk 4]
```diff
@@ -172,28 +174,74 @@ class storageReport extends viewModelBase {
...
+                    lazyLoadChildren: false,
+                    type: sampleItem.type,
+                    parent: sampleItem.parent,
+                    showType: sampleItem.showType,
+                    formatSize: sampleItem.formatSize,
+                    formatPercentage: sampleItem.formatPercentage,
+                    recyclableJournal: sampleItem.recyclableJournal,
+                    customSizeProvider: sampleItem.customSizeProvider,
+                    isStorageEnvironment: sampleItem.isStorageEnvironment
+                })
```
Comment (ml054) [inline src/Raven.Studio/typescript/viewmodels/database/status/storageReport.ts]: this looks very suspicious - what do you want to achieve here?

File: src/Raven.Studio/typescript/models/database/status/storageReportItem.ts [hunk 5]
```diff
@@ -19,15 +20,18 @@ class storageReportItem {
     numberOfEntries: number = null;
     lazyLoadChildren = false;
     customSizeProvider: (header: boolean) => string;
-
+    isGrouped: boolean;
+    
     recyclableJournal = false;
 
-    constructor(name: string, type: string, showType: boolean, size: number, internalChildren: storageReportItem[] = null) {
+    constructor(name: string, type: string, showType: boolean, size: number, length: number, internalChildren: storageReportItem[] = null, isGrouped: boolean = false) {
```
Comment (ml054) [inline src/Raven.Studio/typescript/models/database/status/storageReportItem.ts]: if we need length param but it is null 90% of use cases then don't put that as constructor param - set this using property - only when needed.

Comment (Danielle9897) [reply src/Raven.Studio/typescript/models/database/status/storageReportItem.ts]: done

File: src/Raven.Studio/typescript/models/database/status/storageReportItem.ts (hunk 1 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/models/database/status/storageReportItem.ts:11]: Length is used to calculate the length for '_stream_'
Needed for the `customSizeProvider` (inside method` mapStreams()`)

File: src/Raven.Studio/typescript/viewmodels/database/status/storageReport.ts (hunk 3 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/viewmodels/database/status/storageReport.ts]: changed to `_.clone()`

File: src/Raven.Studio/typescript/viewmodels/database/status/storageReport.ts (hunk 4 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/viewmodels/database/status/storageReport.ts]: just needed to provide an initial `storageReportItem `object for the reduce

File: src/Raven.Studio/typescript/models/database/status/storageReportItem.ts (hunk 1 above)
Comment (ml054) [reply src/Raven.Studio/typescript/models/database/status/storageReportItem.ts:11]: Can't we use size?

File: src/Raven.Studio/typescript/viewmodels/database/status/storageReport.ts (hunk 3 above)
Comment (ml054) [reply src/Raven.Studio/typescript/viewmodels/database/status/storageReport.ts]: Can we avoid clone at all?

File: src/Raven.Studio/typescript/models/database/status/storageReportItem.ts (hunk 1 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/models/database/status/storageReportItem.ts:11]: For streams, the server returns both `Length` & `AllocatedSpaceInBytes`
which is used here:
https://github.com/ravendb/ravendb/blob/7f32742ee1fe90c30b9d4242cdb3234a3b589268/src/Raven.Studio/typescript/viewmodels/database/status/storageReport.ts#L225

File: src/Raven.Studio/typescript/viewmodels/database/status/storageReport.ts (hunk 3 above)
Comment (Danielle9897) [reply src/Raven.Studio/typescript/viewmodels/database/status/storageReport.ts]: yes - removed

**Review by ml054 — APPROVED** [review]

**Pull request merged**